tools/src/main/java/org/apache/kafka/tools/ConnectPluginPath.java

```diff
@@ -0,0 +1,399 @@
+    // visible for testing
+    static Set<String> parse(URL u) {
+        Set<String> names = new LinkedHashSet<>(); // preserve insertion order
+        try {
+            URLConnection uc = u.openConnection();
+            uc.setUseCaches(false);
+            try (InputStream in = uc.getInputStream();
+                 BufferedReader r
+                         = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
+                int lc = 1;
+                while ((lc = parseLine(r, lc, names)) >= 0) {
+                    // pass
+                }
+            }
+        } catch (IOException x) {
+            throw new RuntimeException("Error accessing ServiceLoader manifest file " + u, x);
+        }
+        return names;
+    }
+
+    private static int parseLine(BufferedReader r, int lc, Set<String> names) throws IOException {
+        String ln = r.readLine();
+        if (ln == null) {
+            return -1;
+        }
+        int ci = ln.indexOf('#');
+        if (ci >= 0) ln = ln.substring(0, ci);
+        ln = ln.trim();
+        int n = ln.length();
+        if (n != 0) {
+            if ((ln.indexOf(' ') >= 0) || (ln.indexOf('\t') >= 0))
+                throw new IOException("Illegal configuration-file syntax");
+            int cp = ln.codePointAt(0);
+            if (!Character.isJavaIdentifierStart(cp))
+                throw new IOException("Illegal provider-class name: " + ln);
+            int start = Character.charCount(cp);
+            for (int i = start; i < n; i += Character.charCount(cp)) {
+                cp = ln.codePointAt(i);
+                if (!Character.isJavaIdentifierPart(cp) && (cp != '.'))
+                    throw new IOException("Illegal provider-class name: " + ln);
+            }
+            names.add(ln);
+        }
+        return lc + 1;
+    }
+
+    private static Map<String, Set<ManifestEntry>> indexManifests(Set<ManifestEntry> manifests) {
+        Map<String, Set<ManifestEntry>> classToManifests = new HashMap<>();
+        for (ManifestEntry manifestEntry : manifests) {
+            Set<ManifestEntry> uris = classToManifests.computeIfAbsent(manifestEntry.className, ignored -> new HashSet<>());
+            uris.add(manifestEntry);
+        }
+        return classToManifests;
+    }
+
+    private static Map<String, List<String>> invertAliases(Map<String, String> aliases) {
+        return aliases.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getValue, e -> Collections.singletonList(e.getKey()), (a, b) -> {
+                    if (a instanceof ArrayList) {
+                        a.addAll(b);
+                        return a;
+                    } else {
+                        List<String> result = new ArrayList<>();
+                        result.addAll(a);
+                        result.addAll(b);
+                        return result;
+                    }
+                }));
```

Comment:
Nit: we can simplify this a bit by always using an `ArrayList`:
```suggestion
                .collect(Collectors.toMap(
                        Map.Entry::getValue,
                        e -> new ArrayList<>(Collections.singleton(e.getKey())),
                        (v1, v2) -> {
                            v1.addAll(v2);
                            return v1;
                        }
                ));
```